libpod/runtime_pod_pause_linux.go

```diff
@@ -0,0 +1,234 @@
+// +build linux
+
+package libpod
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/containers/image/storage"
+	"github.com/containers/image/transports/alltransports"
+	"github.com/containers/image/types"
+	"github.com/containers/storage/pkg/archive"
+	digest "github.com/opencontainers/go-digest"
+	"github.com/opencontainers/runtime-tools/generate"
+	"github.com/pkg/errors"
+	"github.com/projectatomic/buildah"
+	"github.com/projectatomic/buildah/util"
+	"github.com/projectatomic/libpod/libpod/image"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	pauseCmd      = "/pause"
+	pauseLocation = "bin/pause"
+	pauseImgName  = "pause-infra-container"
+)
+
+func (r *Runtime) buildPauseImage(ctx context.Context, p *Pod) (string, error) {
+	// Buildah new
+	// TODO make sure this matches global cli.context
+	signaturePolicy := ""
+	store := r.store
+	commonOpts := &buildah.CommonBuildOptions{}
+	transport := DefaultTransport
+	isolation := buildah.IsolationDefault
+	sc := &types.SystemContext{}
+	pullPolicy := buildah.PullNever
+	namespaceOptions := make(buildah.NamespaceOptions, 0, 7)
+
+	builderOptions := buildah.BuilderOptions{
+		FromImage:             "scratch",
+		Transport:             transport,
+		Container:             "pause",
+		PullPolicy:            pullPolicy,
+		SignaturePolicyPath:   signaturePolicy,
+		SystemContext:         sc,
+		DefaultMountsFilePath: "",
+		Isolation:             isolation,
+		NamespaceOptions:      namespaceOptions,
+		CommonBuildOpts:       commonOpts,
+	}
+
+	builderOptions.ReportWriter = os.Stderr
+
+	builder, err := buildah.NewBuilder(ctx, store, builderOptions)
+	if err != nil {
+		return "", err
+	}
+
+	// Buildah onBuild
+	addOptions := buildah.AddAndCopyOptions{
+		Chown:  "",
+		Hasher: digest.Canonical.Digester().Hash(),
+	}
+
+	err = builder.Add(pauseCmd, false, addOptions, pauseLocation)
+	if err != nil {
+		return "", err
+	}
+
+	builder.SetEntrypoint([]string{pauseCmd})
+	builder.SetCmd([]string{pauseCmd})
+
+	// Buildah commit
+	image := fmt.Sprintf("localhost/%s", pauseImgName)
+	compress := archive.Gzip
```

Comment:
Why are we building/committing our own pause image? Can we just pull the same one CRI-O uses?

I'd let the exact image be configured in the config file, but I don't think we need to build a fresh one

Reply:
So I based a lot of this on how CRI-O does it. They initially pulled down the kubernetes image, but through discussion in: https://github.com/kubernetes-incubator/cri-o/issues/91 they opted to build it themselves in: https://github.com/kubernetes-incubator/cri-o/pull/93 . I decided to go that route early to not have to remake the same decision they did.

Reply:
They don't procedurally build it, though - the default pause image in CRI-O is pulled (they don't have the ability to build images). They are currently defaulting to https://github.com/kubernetes-incubator/cri-o/blob/master/lib/config.go#L17

Reply:
I personally would prefer building it ourselves, but it does simplify the code a lot so I'll remove it.